src/main/java/org/openstreetmap/atlas/geography/atlas/change/FeatureChange.java

```diff
@@ -246,131 +283,241 @@ public String toJson()
         }
-        return FeatureChange.add(result);
+        return this;
     }
 
-    private <M> M mergedMember(final String memberName, final AtlasEntity left,
-            final AtlasEntity right, final Function<AtlasEntity, M> memberExtractor,
-            final Optional<BinaryOperator<M>> memberMergerOption)
+    /**
+     * Compute the beforeView using a given afterView and Atlas context. The beforeView is a
+     * CompleteEntity, and will contain only those fields which have been updated in the afterView.
+     * E.g. Suppose the afterView is a CompleteNode with ID 1, and the only updated field is the tag
+     * Map. Then the beforeView will be a CompleteNode with ID 1, where the only populated field is
+     * the tag Map. However, the afterView's tag Map will be fetched from the Atlas context. Having
+     * a beforeView for each {@link FeatureChange} allows for more robust merging strategies.
+     *
+     * @param atlas
+     *            the atlas context
+     */
+    private void computeBeforeViewUsingAtlasContext(final Atlas atlas) // NOSONAR
     {
-        final M result;
-        final M leftMember = memberExtractor.apply(left);
-        final M rightMember = memberExtractor.apply(right);
-        if (leftMember != null && rightMember != null)
+        if (atlas == null)
         {
-            // Both are not null, merge evaluated
-            if (leftMember.equals(rightMember))
-            {
-                // They are equal, arbitrarily pick one.
-                result = leftMember;
-            }
-            else if (memberMergerOption.isPresent())
-            {
-                // They are unequal, but we can attempt a merge
-                try
+            throw new CoreException("Atlas context cannot be null");
+        }
+
+        AtlasEntity beforeViewUpdatesOnly;
+        final AtlasEntity beforeViewFromAtlas = atlas.entity(this.updatedView.getIdentifier(),
+                this.updatedView.getType());
+        if (beforeViewFromAtlas == null)
+        {
+            throw new CoreException("Could not find {} with ID {} in atlas context",
+                    this.updatedView.getType(), this.updatedView.getIdentifier());
+        }
+
+        /*
+         * Make type specific updates first.
+         */
+        switch (this.updatedView.getType())
+        {
+            /*
+             * Area specific updates. The only Area-specific field is the polygon.
+             */
+            case AREA:
+                final Area updatedAreaView = (Area) this.updatedView;
+                final Area beforeAreaView = (Area) beforeViewFromAtlas;
+                beforeViewUpdatesOnly = CompleteArea.shallowFrom(beforeAreaView);
+                if (updatedAreaView.asPolygon() != null)
                 {
-                    result = memberMergerOption.get().apply(leftMember, rightMember);
+                    ((CompleteArea) beforeViewUpdatesOnly).withPolygon(beforeAreaView.asPolygon());
```

Comment:
Yes, because `beforeViewUpdatesOnly` is of type `AtlasEntity`. So I need to safely downcast in the switch statement to access the right specialized subclass methods (`withPolygon` in this case).